### PR TITLE
CB-20500 Freeipa downscale failed from 4 nodes

### PR DESCRIPTION
**Pull request by gergopapi2**
During freeipa upgrade an error happened, because of which 4 nodes remained. Using downscale from 4 nodes was not possible, however, as mapping 4 nodes to AvailabilityType was not possible, as the 4 node state was meant to be temporary.

The present commit thus adds an error handling scenario: it introduces a new state, TEMPORARY_FOUR_NODE_BASED, for the case when the freeipa ends up having 4 nodes. It is now possible to downscale to HA. A downscale to TWO_NODE_BASED system is not allowed, as this is an error fixing scenario and let's allow only the smallests change to the system. Once the customer is back to HA, and also replication agreements are ok, he can separately decide to go downscale to 2 nodes.

The present commit also adds numerous log lines to repair.

See detailed description in the commit message.

### CODE REVIEW

File: freeipa-api/src/main/java/com/sequenceiq/freeipa/api/v1/freeipa/stack/model/common/AvailabilityType.java [hunk 1]
```diff
@@ -8,6 +8,7 @@
 
 public enum AvailabilityType {
 
+    TEMPORARY_FOUR_NODE_BASED(4),
```
Comment (lacikaaa) [inline freeipa-api/src/main/java/com/sequenceiq/freeipa/api/v1/freeipa/stack/model/common/AvailabilityType.java:11]: while this should work for now I think this is a bit fragile. What happens if we have 5 instance somehow? I think we should have some generic solution to handle the case which is an unsupported scenario. And if we start supporting 4 node based installation then it would be a new valid config here while the error handling would be the same

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/RepairInstancesService.java [hunk 2]
```diff
@@ -259,29 +262,27 @@ public OperationStatus rebootInstances(String accountId, RebootInstancesRequest
                 getInstancesToRepair(healthMap, allInstancesByInstanceId, request.getInstanceIds(), request.isForceReboot(), true);
 
         if (instancesToReboot.keySet().isEmpty()) {
-            throw new NotFoundException("No unhealthy instances to reboot.  Maybe use the force option.");
+            throwNotFoundException("No unhealthy instances to reboot. Maybe use the force option.");
```
Comment (pdarvasi) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/RepairInstancesService.java:265]: same

File: freeipa-api/src/main/java/com/sequenceiq/freeipa/api/v1/freeipa/stack/model/common/AvailabilityType.java (hunk 1 above)
Comment (pdarvasi) [reply freeipa-api/src/main/java/com/sequenceiq/freeipa/api/v1/freeipa/stack/model/common/AvailabilityType.java:11]: just an idea as it would require some more code modification and mappings but something like `TEMPORARY(0)` could work

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaScalingValidationService.java [hunk 3]
```diff
@@ -46,13 +43,26 @@ public void validateStackForVerticalUpscale(Stack stack, VerticalScaleRequest re
...
-        if (allInstances.size() < AvailabilityType.HA.getInstanceCount()) {
-            LOGGER.warn("FreeIPA instance count is not allowed. Size: [{}]", allInstances.size());
-            throw new BadRequestException("Downscaling currently only available for FreeIPA installation with 3 instances");
-        }
+        validateScalingIsDownscale(scalingPath);
         executeCommonValidations(allInstances, stack, scalingPath, OperationType.DOWNSCALE);
     }
 
+    private void validateScalingIsUpscale(ScalingPath scalingPath) {
+        if (scalingPath.getOriginalAvailabilityType().getInstanceCount() > scalingPath.getTargetAvailabilityType().getInstanceCount()) {
```
Comment (pdarvasi) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaScalingValidationService.java:51]: shouldn't we compare with the actual original instance count here?

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaScalingValidationService.java [hunk 4]
```diff
@@ -46,13 +43,26 @@ public void validateStackForVerticalUpscale(Stack stack, VerticalScaleRequest re
...
+            throw new BadRequestException(String.format("Refusing %s as target node count is smaller than current. Current node count: %d, " +
+                            "target node count: %d.", OperationType.UPSCALE.getLowerCaseName(), scalingPath.getOriginalAvailabilityType().getInstanceCount(),
+                    scalingPath.getTargetAvailabilityType().getInstanceCount()));
+        }
+    }
+
+    private void validateScalingIsDownscale(ScalingPath scalingPath) {
+        if (scalingPath.getOriginalAvailabilityType().getInstanceCount() < scalingPath.getTargetAvailabilityType().getInstanceCount()) {
```
Comment (pdarvasi) [inline freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaScalingValidationService.java:59]: shouldn't we compare with the actual original instance count here?

File: freeipa/src/main/resources/application.yml [hunk 5]
```diff
@@ -142,7 +142,9 @@ freeipa:
         - TWO_NODE_BASED
       TWO_NODE_BASED:
         - HA
+        - TWO_NODE_BASED
```
Comment (pdarvasi) [inline freeipa/src/main/resources/application.yml:145]: why is this needed?

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaScalingValidationService.java (hunk 3 above)
Comment (gergopapi2) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaScalingValidationService.java:51]: There are 2 separate validations.
This validation only checks that the availability type is not going to decrease (i.e. the user sent a downscale request to the upscale endpoint). So an HA => HA scaling is also allowed.
There is another validation (in the common part) where the instance count is checked if it will change at all. If not, a validation error is thrown.

File: freeipa/src/main/resources/application.yml (hunk 5 above)
Comment (gergopapi2) [reply freeipa/src/main/resources/application.yml:145]: Currently it is not needed, indeed, as you do not have means to distinguish a correctly running HA cluster from a TWO_NODE_BASED one where because of a failure in rolling upgrade has only 3 nodes. 
I will thus remove it and put back once AvailabilityType is also saved to the DB.

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaScalingValidationService.java (hunk 4 above)
Comment (gergopapi2) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaScalingValidationService.java:59]: This is just making sure that it is not an upscale request that was sent to the downscale endpoint.

File: freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaScalingValidationService.java (hunk 3 above)
Comment (pdarvasi) [reply freeipa/src/main/java/com/sequenceiq/freeipa/service/stack/FreeIpaScalingValidationService.java:51]: makes sense, thank you

**Review by pdarvasi — APPROVED** [review]
